maven-core/src/main/java/org/apache/maven/graph/DefaultGraphBuilder.java

```diff
@@ -401,18 +403,49 @@ else if ( selectedProject.isDirectory() )
             return projects;
         }
 
-        List<File> files = Arrays.asList( request.getPom().getAbsoluteFile() );
+        List<File> files = Collections.singletonList( request.getPom().getAbsoluteFile() );
         collectProjects( projects, files, request );
         return projects;
     }
 
-    private void collectProjects( List<MavenProject> projects, List<File> files, MavenExecutionRequest request )
-        throws ProjectBuildingException
+    private void collectProjects( List<MavenProject> projects, final List<File> files,
+                                  final MavenExecutionRequest request ) throws ProjectBuildingException
     {
-        ProjectBuildingRequest projectBuildingRequest = request.getProjectBuildingRequest();
+        final ProjectBuildingRequest projectBuildingRequest = request.getProjectBuildingRequest();
+
+        List<ProjectBuildingResult> results;
+        long start = System.currentTimeMillis();
+        if ( request.getDegreeOfConcurrency() <= 1 || Boolean.getBoolean( "maven.concurrentGraph.disable" ) )
```

Comment:
Why binding this improvement to **request.getDegreeOfConcurrency()**?

This is only about building the model and not executing plugins, isn't it?

Reply:
Hi
in any case need to cap resources allocated at this step as build owner needs to control his environment. if user set -T flag we can assume he gave consent that build can use up to those CPU resources . So using DegreeOfConcurrency sounds natural for me. To make it more fine grained it is possible to introduce additional control flags, but i decided to not complicate configuration at this point